File: c_files/raylib/dungeon_delve_text_rpg/source/menu.c
#include "../include/menu.h"
#include <raylib.h>
#include <stdio.h>
#include <stdlib.h>

void
menu_init(Menu* menu,
		  const char* arr[],
		  int arr_size,
		  int font_size,
		  int line_gap,
		  int margin,
		  Rectangle layout_rect)
{
	menu->arr_size = arr_size;
	menu->selected = 0;
	menu->first_visible_element = 0;
	menu->font_size = font_size;
	menu->line_gap = line_gap;
	menu->line_height = font_size + line_gap;
	menu->margin = margin;
	menu->layout_rect = layout_rect;
	menu->max_visible_elements =
	  (layout_rect.height - line_gap - margin * 2) / menu->line_height;
	menu->visible_text_height =
	  (font_size * arr_size) + (line_gap * (arr_size - 1));

	menu->items = malloc(arr_size * sizeof(MenuItem));
	if (menu->items == NULL) {
		fprintf(stderr, "Memory not allocated to menu->items\n");
		exit(EXIT_FAILURE);
	}

	for (int i = 0; i < arr_size; i++) {
		menu->items[i].title = arr[i];
		menu->items[i].width = MeasureText(arr[i], font_size);
		menu->items[i].x = (int)layout_rect.x + (int)layout_rect.width / 2 -
						   menu->items[i].width / 2;
	}
}

void
menu_draw(Menu* menu)
{
	int last_visible_element =
	  menu->first_visible_element + menu->max_visible_elements - 1;

	for (int i = menu->first_visible_element;
		 i < menu->arr_size && i < last_visible_element;
		 i++) {
		menu->items[i].y =
		  (int)menu->layout_rect.y + (int)menu->layout_rect.height / 2 -
		  menu->visible_text_height / 2 +
		  (i - menu->first_visible_element) * menu->line_height;

		// draw items
		Color color = (i == menu->selected) ? BLACK : GRAY;

		DrawText(menu->items[i].title,
				 menu->items[i].x,
				 menu->items[i].y,
				 menu->font_size,
				 color);

		// draw selection arrow
		if (i == menu->selected) {
			int arrow_margin = 30;
			DrawText(">",
					 menu->items[i].x - arrow_margin,
					 menu->items[i].y,
					 menu->font_size,
					 color);
		}
	}
}

void
menu_update(Menu* menu)
{
	if (IsKeyPressed(KEY_UP)) {
		menu->selected--;
		if (menu->selected <= 0) {
			menu->selected = 0;
		}

		if (menu->selected < menu->first_visible_element) {
			menu->first_visible_element = menu->selected;
		}
	}
	if (IsKeyPressed(KEY_DOWN)) {
		menu->selected++;
		if (menu->selected > menu->arr_size - 1) {
			menu->selected = menu->arr_size - 1;
		}

		if (menu->selected >
			menu->first_visible_element + menu->max_visible_elements - 1) {
			menu->first_visible_element++;
		}
	}
}

void
menu_free(Menu* menu)
{
	free(menu->items);
	menu->items = NULL;
}

File: python_files/raylib/text_rpg/menu.py
import pyray as p


class Menu:
    def __init__(
        self,
        text_list: list[str],
        font_size: int,
        text_gap: int,
        layout_rect: p.Rectangle,
    ) -> None:
        self.text_list: list[str] = text_list
        self.font_size: int = font_size
        self.text_gap: int = text_gap
        self.text_height: int = self.font_size + self.text_gap
        self.layout_rect: p.Rectangle = layout_rect

        self.selected_element_idx: int = 0
        self.first_visible_element_idx: int = 0
        self.max_visible_elements: int = int(
            (self.layout_rect.height - self.text_gap) / self.text_height
        )
        self.max_visible_height: int = (self.font_size * len(self.text_list)) + (
            self.text_gap * (len(self.text_list) - 1)
        )

    def draw(self) -> None:
        last_visible_element: int = (
            self.first_visible_element_idx + self.max_visible_elements
        )

        for i in range(self.first_visible_element_idx, last_visible_element):
            if i >= 0 and i < len(self.text_list):
                text_width: int = p.measure_text(self.text_list[i], self.font_size)
                text_x: int = int(
                    self.layout_rect.x + self.layout_rect.width / 2 - text_width / 2
                )
                text_y: int = int(
                    self.layout_rect.y
                    + self.layout_rect.height / 2
                    - self.max_visible_height / 2
                    + (i - self.first_visible_element_idx) * self.text_height
                )

                self.color: p.Color = (
                    p.BLACK if i == self.selected_element_idx else p.GRAY
                )

                p.draw_text(
                    self.text_list[i], text_x, text_y, self.font_size, self.color
                )

                if i == self.selected_element_idx:
                    horizontal_gap: int = 30
                    p.draw_text(
                        ">", text_x - horizontal_gap, text_y, self.font_size, self.color
                    )

    def update(self) -> None:
        if p.is_key_pressed(p.KeyboardKey.KEY_UP):
            self.selected_element_idx = (
                self.selected_element_idx - 1 + len(self.text_list)
            ) % len(self.text_list)

        if p.is_key_pressed(p.KeyboardKey.KEY_DOWN):
            self.selected_element_idx = (self.selected_element_idx + 1) % len(
                self.text_list
            )

